[UX] New dashboard & refactored KeyVisual

niels9001 opened the pull request:
## Summary of the Pull Request

### New dashboard UX

https://github.com/user-attachments/assets/6d5e37f4-d071-488b-bb86-b1fa16859416

**Update available:**
![image](https://github.com/user-attachments/assets/64a2406d-4493-47cc-b86c-59ed2209a4a9)

**No update available:**
![image](https://github.com/user-attachments/assets/02cf2d5f-c235-4544-ac47-09aad85434ef)

### Updated `KeyVisual` and `Shortcut` control
- Refactoring `KeyVisual` to remove redundant properties and UI elements, and using Styles for better customization.
- Shortcut control now shows a "Configure shortcut" label when there's no shortcut configured.
![image](https://github.com/user-attachments/assets/fe3cb4b6-07e8-4714-8359-47dac2bccca5)

### Other changes
- Consolidated converters that were used across pages in `App.xaml.cs` with consistent naming.
- Renamed templated controls (from `.cs` to `.xaml.cs`) and moving those to the `Controls` root folder vs. individual folders for a better overview.
<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #39520, #32944
- [ ] **Communication:** I've discussed this with core contributors already. If the work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end-user-facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed, or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

niels9001 committed:
Cleaning up Controls folder

niels9001 committed:
Adding Card control

niels9001 committed:
Dashboard changes

niels9001 committed:
Remove KBM dashboard item and related code

Eliminated the DashboardModuleKBMItem, its template, and associated logic from the dashboard. This simplifies the DashboardViewModel and DashboardPage.xaml by removing the display and dynamic updating of Keyboard Manager remappings.

niels9001 committed:
Adding shortcuts view

niels9001 committed:
More shortcuts

niels9001 committed:
Refactor KeyVisual for better UX customization

niels9001 committed:
Updating shortcut strings

niels9001 committed:
Enabling resizing

niels9001 committed:
Adding icons to the quick actions

niels9001 committed:
CI fixes

niels9001 committed:
More shortcut visual refactoring

niels9001 committed:
More changes

niels9001 committed:
Merge branch 'main' into niels9001/new-dashboard-ux

niels9001 committed:
Add missing tag

niels9001 committed:
Moving Update and ShortcutConflicts to dedicated controls so they can be reused

niels9001 committed:
Fix

niels9001 committed:
Bringing back the icon for the shortcut control

niels9001 committed:
Styles refactor

niels9001 committed:
Update SettingsPageControl.xaml

niels9001 committed:
Merge branch 'main' into niels9001/new-dashboard-ux

niels9001 committed:
Update App.xaml

crutkas commented:
/azp run

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/Dashboard/CheckUpdateControl.xaml:
```diff
@@ -0,0 +1,80 @@
+                <Border
+                    Width="20"
+                    Height="20"
+                    CornerRadius="10">
```
Why not `OverlayCornerRadius`?

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/Dashboard/CheckUpdateControl.xaml:
The two Grid containers are basically the same, just with different content. Is it an idea to use as much code as possible a single time?

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/Dashboard/ShortcutConflictControl.xaml:
```diff
@@ -0,0 +1,33 @@
+                    <TextBlock
+                        Foreground="{ThemeResource TextFillColorSecondaryBrush}"
+                        Style="{StaticResource CaptionTextBlockStyle}"
+                        Text="2 conflicts found" />
```
Note: hard-coded text

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/KeyVisual.xaml:
```diff
@@ -0,0 +1,189 @@
+        <Setter Property="HorizontalContentAlignment" Value="Center" />
+        <Setter Property="VerticalContentAlignment" Value="Center" />
+        <Setter Property="VerticalAlignment" Value="Center" />
+        <Setter Property="FontSize" Value="{StaticResource ControlContentThemeFontSize}" />
```
Line 17 says "12"

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/ShortcutControl/ShortcutDialogContentControl.xaml:
```diff
@@ -35,11 +35,15 @@
                         Content="{Binding}"
-                        IsError="{Binding ElementName=ShortcutContentControl, Path=IsError, Mode=OneWay}"
+                        CornerRadius="{StaticResource ControlCornerRadius}"
+                        FontSize="19"
+                        FontWeight="SemiBold"
```
Or `SubtitleTextBlockStyle`?

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Views/MouseUtilsPage.xaml:
```diff
@@ -153,7 +149,7 @@
                         IsOpen="True"
                         Severity="Informational"
-                        Visibility="{x:Bind ViewModel.IsAnimationEnabledBySystem, Mode=OneWay, Converter={StaticResource BoolToInvertedVisibilityConverter}}">
+                        Visibility="{x:Bind ViewModel.IsAnimationEnabledBySystem, Mode=OneWay, Converter={StaticResource BoolToReverseVisibilityConverter}}">
```
Is this the same as `ReverseBoolToVisibilityConverter`?

Jay-o-Way commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -3205,18 +3211,21 @@ Activate by holding the key for the character you want to add an accent to, then
   </data>
   <data name="ShortcutGuide_ShortDescription" xml:space="preserve">
-    <value>Show a help overlay with Windows shortcuts</value>
+    <value>Launch Shortcut Guide</value>
```
```suggestion
    <value>Open Shortcut Guide</value>
```

>Don't use to mean opening an app or form. Use _open_ instead.

https://learn.microsoft.com/nl-nl/style-guide/a-z-word-list-term-collections/l/launch

Jay-o-Way commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -3192,10 +3198,10 @@ Activate by holding the key for the character you want to add an accent to, then
   </data>
   <data name="Run_ShortDescription" xml:space="preserve">
-    <value>A quick launcher</value>
+    <value>Launch Run</value>
```
```suggestion
    <value>Open Run</value>
```

>Don't use to mean opening an app or form. Use _open_ instead.

https://learn.microsoft.com/nl-nl/style-guide/a-z-word-list-term-collections/l/launch

Jay-o-Way commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -4224,10 +4236,10 @@ Activate by holding the key for the character you want to add an accent to, then
   </data>
   <data name="Shell_Dashboard.Content" xml:space="preserve">
-    <value>Dashboard</value>
+    <value>Home</value>
```
But what if user is at their work..?
Kidding 😆

niels9001 committed:
Update src/settings-ui/Settings.UI/Strings/en-us/Resources.resw

Co-authored-by: Jay <65828559+Jay-o-Way@users.noreply.github.com>

niels9001 committed:
Update src/settings-ui/Settings.UI/Strings/en-us/Resources.resw

Co-authored-by: Jay <65828559+Jay-o-Way@users.noreply.github.com>

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Views/DashboardPage.xaml:
```diff
@@ -15,491 +15,349 @@
+                                    <DataTemplate x:DataType="viewmodels:DashboardListItem">
+                                        <Grid>
+                                            <ItemsControl
+                                                Grid.Column="1"
```
Why a **Column**, if there are no column definitions? :)

Jaylyn-Barbee replied:
Removed.

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/KeyVisual.xaml.cs:
```diff
@@ -107,7 +105,7 @@ private void Update()
-                            _keyVisual._keyPresenter.Content = "\uE752"; break; */
+                            _keyVisual._keyPresenter.Content = "\uE752"; break;
 
                         case 38: _keyVisual._keyPresenter.Content = "\uE0E4"; break; // The Up Arrow key or button.
```
Would you mind creating suggestions for the glyphs that are using unknown/outdated glyphs that I could commit directly?

Jaylyn-Barbee replied:
Note to self: check if this fixes accessibility issue.

Jaylyn-Barbee replied:
Using any symbols raises accessibility issues since we are using "Unicode characters from the private range" when these get set as the "name" of the textbox

niels9001 replied:
> Using any symbols raises accessibility issues since we are using "Unicode characters from the private range" when these get set as the "name" of the textbox

Could we set the AutomationProperties here? Or create a string that returns the name vs the glyph as an additional properties?

Jaylyn-Barbee replied:
I was able to wrap the content in a text block and give that text block an accessible name.

niels9001 committed:
Update KeyVisual.xaml

Jaylyn-Barbee committed:
Fixed a few accessibility issues. Improper name fields and such

Jaylyn-Barbee committed:
Merge branch 'main' of https://github.com/Jaylyn-Barbee/PowerToys into niels9001/new-dashboard-ux

Jaylyn-Barbee committed:
merging main

Jaylyn-Barbee committed:
testing

Jaylyn-Barbee committed:
Recovered work

Jaylyn-Barbee committed:
spelling issue + xaml formatting

Jaylyn-Barbee committed:
ran xaml formatter on unfomratted file

Jaylyn-Barbee committed:
Replace the word 'launch' with 'open' since we shouldn't use 'launch' to describe opening something per guidelines

Jaylyn-Barbee committed:
fixed a11y issue with glyphs in keyvisual + removed unneeded line

Jaylyn-Barbee commented:
There is one last a11y issue that will be fixed when this is merged and we update https://github.com/CommunityToolkit/Windows/pull/702

Jaylyn-Barbee committed:
Fixed dashboard page styling

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/KeyVisual.xaml.cs:
```diff
@@ -0,0 +1,210 @@
+                        accessibleName = "Windows key";
+                        PathIcon winIcon = XamlReader.Load(
+                            @"<PathIcon xmlns=""http://schemas.microsoft.com/winfx/2006/xaml/presentation"" 
+                    Data=""M896 896H0V0h896v896zm1024 0h-896V0h896v896zM896 1920H0v-896h896v896zm1024 0h-896v-896h896v896z"" />") as PathIcon;
```
Oh, TIL! Yeah, not a big fan of the XAMLReader implementation anyway.

For consistency, should we define the path data in App.xaml and refer to that? We might want to use the Windows logo in other places too.

Jaylyn-Barbee replied:
I can try to implement this today. I will report back when I need someone to review the implementation!

ghost1372 replied:
@Jaylyn-Barbee , i have KeyVisual in my Library, and i use PathIcon, i dont see any issues!
<img width="205" height="75" alt="image" src="https://github.com/user-attachments/assets/5df6a545-5942-44db-b862-2dee7e710562" />

Jaylyn-Barbee committed:
fixed AOT issue with windows logo in keyvisual

Jaylyn-Barbee committed:
xaml formatting

Jaylyn-Barbee committed:
Added back string for 'Update Available'

Jaylyn-Barbee committed:
fixed wrap layout issue

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/SettingsPageControl/SettingsPageControl.xaml:
```diff
@@ -9,17 +9,10 @@
     <UserControl.Resources>
         <x:Double x:Key="PageMaxWidth">1000</x:Double>
         <x:Double x:Key="PageHeaderMaxWidth">1020</x:Double>
-        <tkconverters:DoubleToVisibilityConverter
```
Great catch

niels9001 commented on src/settings-ui/Settings.UI/SettingsXAML/Views/DashboardPage.xaml.cs:
```diff
@@ -55,5 +54,19 @@ private void DashboardListItemClick(object sender, RoutedEventArgs e)
+        {
+            if (App.GetOobeWindow() == null)
+            {
+                App.SetOobeWindow(new OobeWindow(Microsoft.PowerToys.Settings.UI.OOBE.Enums.PowerToysModules.WhatsNew));
+            }
+            else
+            {
+                App.GetOobeWindow().SetAppWindow(OOBE.Enums.PowerToysModules.WhatsNew);
```
Good catch! Created a global namespace at the top of the file.

niels9001 committed:
Address comments

niels9001 committed:
Cleaning up and loccing KeyVisual

Jay-o-Way commented on src/settings-ui/Settings.UI/SettingsXAML/Controls/KeyVisual.xaml.cs:
```diff
@@ -210,7 +212,7 @@ private TextBlock CreateGlyphTextBlock(string glyph, string accessibleName)
             {
                 Text = glyph,
                 FontFamily = new FontFamily("Segoe Fluent Icons, Segoe MDL2 Assets"),
```
Can here be used a variable?

niels9001 committed:
Refactoring KeyVisual (#40904)

We had 2 issues with the refactoring of KeyVisual that happened in
#40214:

- We had little control over the pixel perfection of the Windows + text
+ glyph KeyVisual templates, which resulted into subpar vertical
alignments.
- When changing theme, the Windows key foreground brush was not
updating.

With this PR
- a new control with dedicated templates is introduced; this allows us
to have full control over the individual templates, and doing all of the
styling in XAML vs in C#.
- I've removed setting the accessible names because those should be set
on the parent container, as Narrator is supposed to announce these
shortcuts in its totality vs individual keys.
- A new property `RenderKeyAsGlyph` has been introduced to render a key
as an icon (to save space in certain scenarios) or to show the full text
(default)

<img width="133" height="147" alt="image"
src="https://github.com/user-attachments/assets/66a342da-fdcf-4c97-80b6-93b9955ebdd3"
/>

<img width="242" height="58" alt="image"
src="https://github.com/user-attachments/assets/0b41cfee-8a7d-409a-83b4-72d13e0b131c"
/>
